BuildingBlocks/test/BuildingBlocks.Application.Tests/Mediatr/QuotaEnforcerBehaviorTests.cs

```diff
@@ -0,0 +1,100 @@
+﻿using Enmeshed.BuildingBlocks.Application.Abstractions.Exceptions;
+using Enmeshed.BuildingBlocks.Application.Attributes;
+using Enmeshed.BuildingBlocks.Application.MediatR;
+using Enmeshed.BuildingBlocks.Domain;
+using FluentAssertions;
+using MediatR;
+using Xunit;
+
+namespace Enmeshed.BuildingBlocks.Application.Tests.Mediatr;
+public class QuotaEnforcerBehaviorTests
+{
+    public QuotaEnforcerBehaviorTests() { }
+
+    [ApplyQuotasForMetrics("MetricKey")]
+    private class TestCommand : IRequest<IResponse> { }
+
+    private class IResponse { }
+
+    [Fact]
+    public async void Throws_QuotaExhaustedException_OnExhaustedQuota()
+    {
+        // Arrange
+        var metricStatusesStubRepository = new MetricStatusesStubRepository(new() {
+            new MetricStatus(new MetricKey("KeyOne"), DateTime.UtcNow.AddDays(1)) 
+        });
+
+        var userContextStub = new UserContextStub();
+        var behavior = new QuotaEnforcerBehavior<TestCommand, IResponse>(metricStatusesStubRepository, userContextStub);
+        var nextHasRan = false;
+
+        // Act
+        Func<Task> acting = async () => await behavior.Handle(
+             new TestCommand(),
+            () => {
+                nextHasRan = true;
+                return Task.FromResult(new IResponse());
+            },
+            CancellationToken.None);
+
+        // Assert
+        await acting.Should().ThrowAsync<QuotaExhaustedException>();
+        nextHasRan.Should().BeFalse();
+    }
+
+    [Fact]
+    public async void Throws_QuotaExhaustedException_OnExhaustedQuota_WithFarthestQuotaOnData()
+    {
+        // Arrange
+        var expectedMetricStatus = new MetricStatus(new MetricKey("KeyTwo"), DateTime.UtcNow.AddDays(10));
+        var metricStatusesStubRepository = new MetricStatusesStubRepository(new() {
+            new MetricStatus(new MetricKey("KeyOne"), DateTime.UtcNow.AddDays(1)),
+            expectedMetricStatus
+        });
+
+        var userContextStub = new UserContextStub();
+        var behavior = new QuotaEnforcerBehavior<TestCommand, IResponse>(metricStatusesStubRepository, userContextStub);
+        var nextHasRan = false;
+
+        // Act
+        Func<Task> action = async () => await behavior.Handle(
+            new TestCommand(),
+            () => {
+                nextHasRan = true;
+                return Task.FromResult(new IResponse());
+            },
+            CancellationToken.None);
+
+        // Assert
+        var ex = await Assert.ThrowsAsync<QuotaExhaustedException>(action);
+        ex.MetricKey.Should().Be(expectedMetricStatus.MetricKey);
```

Comment:
Don't mix the assertion libraries. Stay with FluentAssertions.

```suggestion
        (await acting.Should().ThrowAsync<QuotaExhaustedException>())
            .Which.MetricKey.Should().Be(expectedMetricStatus.MetricKey);
```

To get rid of the `await`, you can create a custom extension method:

```c#
public static class NonGenericAsyncFunctionAssertionsExtensions
{
    public static ExceptionAssertions<TException> ThrowAsync2<TException>(this NonGenericAsyncFunctionAssertions assertions, string because = "",
        params object[] becauseArgs)
        where TException : Exception
    {
        return assertions.ThrowAsync<TException>(because, becauseArgs).WaitAsync(CancellationToken.None).Result;
    }
}
```
(you have to find a good name for `ThrowAsync2`)

Then your call would look like:
```c#
acting.Should().ThrowAsync2<QuotaExhaustedException>().Which.MetricKey.Should().Be(expectedMetricStatus.MetricKey);
```